src/radical/pilot/umgr/staging_input/default.py

```diff
@@ -325,26 +325,58 @@ def _handle_unit(self, unit, actionables):
         self._prof.prof("create_sandbox_stop", uid=uid)
 
         # Loop over all transfer directives and execute them.
+        removables = list()
         for sd in actionables:
 
             action = sd['action']
             flags  = sd['flags']
             did    = sd['uid']
             src    = sd['source']
             tgt    = sd['target']
+            if action == rpc.TARBALL:
+                # If the staging method is tarball, check initially if the tar 
+                # file exists. If not create it. Add each file in the tarball
+                # and append the directive to the removables.
 
-            self._prof.prof('staging_in_start', uid=uid, msg=did)
+                if not os.path.isfile(uid+'.tar'):
+                    tar_file = tarfile.open(uid+'.tar','w')
+                tar_file.add(src.split('client:///')[1],arcname=tgt.split('unit:///')[1])
+                removables.append(sd)
+                sid = sd.get('uid')
+            elif action == rpc.TRANSFER:
+                self._prof.prof('staging_in_start', uid=uid, msg=did)
 
-            src = complete_url(src, src_context, self._log)
-            tgt = complete_url(tgt, tgt_context, self._log)
+                src = complete_url(src, src_context, self._log)
+                tgt = complete_url(tgt, tgt_context, self._log)
 
-            if rpc.CREATE_PARENTS in flags:
-                copy_flags = rs.filesystem.CREATE_PARENTS
-            else:
-                copy_flags = 0
+                saga_dir.copy(src, tgt, flags=flags)
+                self._prof.prof('staging_in_stop', uid=uid, msg=did)
+
+
+        if os.path.isfile(uid+'.tar'):
+            # if a tarball exists close it and do the transfer. After it is done
+            # delete the tarball. Remove any directive that was use to create the
+            # tarball and add a single one that communicates to the agent that
+            # this type of transfer was executed.
+            tar_file.close()
+
+            self._prof.prof('staging_in_start', uid=uid, msg=did)
```

Comment:
Ah, here it is.  So the transfer ID is then the did of the last staging directive.  Hmm, Lets discuss this, I am not sure what analytics will think if some SDs do not have any profile events at all.  The files get transferred after all...